### ISSUE TITLE
pip install failure

### ISSUE THREAD

**Issue by thorstenhater** [arbor-sim/arbor#1004]
# Installation through pip fails.

Versions used
- Python 3.6.8
- pip 19.0.3
- CMake 3.14.0

No MPI was loaded at this time.

Sanitised log of the transaction
```
$ pip install arbor
Collecting arbor
  Using cached https://files.pythonhosted.org/packages/88/3a/2dfb1af9b3a8b32b227ef734ab152de299cd696bb7b59a710d5d4b232c5e/arbor-0.3.tar.gz
Building wheels for collected packages: arbor
  Building wheel for arbor (setup.py) ... error
  Complete output from command .../bin/python -u -c "import setuptools, tokenize;__file__='/tmp/pip-install-10l4x7qg/arbor/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" bdist_wheel -d /tmp/pip-wheel-27zn4o4p --python-tag cp36:
  running bdist_wheel
  running build
  running build_ext
  Traceback (most recent call last):
    File "<string>", line 1, in <module>
    File "/tmp/pip-install-10l4x7qg/arbor/setup.py", line 153, in <module>
      'Bug Reports': 'https://github.com/arbor-sim/arbor/issues',
    File ".../lib/python3.6/site-packages/setuptools/__init__.py", line 143, in setup
      return distutils.core.setup(**attrs)
    File ".../lib/python3.6/distutils/core.py", line 148, in setup
      dist.run_commands()
    File "...lib/python3.6/distutils/dist.py", line 955, in run_commands
      self.run_command(cmd)
    File ".../lib/python3.6/distutils/dist.py", line 974, in run_command
      cmd_obj.run()
    File ".../lib/python3.6/site-packages/wheel-0.33.1-py3.6.egg/wheel/bdist_wheel.py", line 192, in run
      self.run_command('build')
    File ".../lib/python3.6/distutils/cmd.py", line 313, in run_command
      self.distribution.run_command(command)
    File ".../lib/python3.6/distutils/dist.py", line 974, in run_command
      cmd_obj.run()
    File "...lib/python3.6/distutils/command/build.py", line 135, in run
      self.run_command(cmd_name)
    File ".../lib/python3.6/distutils/cmd.py", line 313, in run_command
      self.distribution.run_command(command)
    File ".../lib/python3.6/distutils/dist.py", line 974, in run_command
      cmd_obj.run()
    File "/tmp/pip-install-10l4x7qg/arbor/setup.py", line 87, in run
      '-DARB_WITH_MPI={}'.format( 'on' if cl_opt_mpi else 'off'),
  NameError: name 'cl_opt_mpi' is not defined
```

**Comment by bcumming** [comment]
What was the command that you used to execute this?
Was it just:
```bash
pip install arbor
```
What if you build in the same environment with an installation option:
```bash
pip --install-option='--vec' arbor
```

**Comment by bcumming** [comment]
Also, can I have the results of
```bash
pip list
```

**Comment by thorstenhater** [comment]
After purging the environment and re-setup. The error persists, but pip version is slightly different. 
```$ pip list
Package                            Version
---------------------------------- --------
alembic                            1.0.8
apipkg                             1.5
appdirs                            1.4.3
argcomplete                        1.9.5
asn1crypto                         0.24.0
atomicwrites                       1.3.0
attrs                              19.1.0
backcall                           0.1.0
backports.shutil-get-terminal-size 1.0.0
bcrypt                             3.1.6
bitstring                          3.1.5
blist                              1.3.6
certifi                            2019.3.9
cffi                               1.12.2
chardet                            3.0.4
Click                              7.0
cloudpickle                        0.8.1
ClusterShell                       1.8
configparser                       3.7.4
cryptography                       2.6.1
Cython                             0.29.6
deap                               1.2.2
decorator                          4.4.0
ecdsa                              0.13
enum34                             1.1.6
execnet                            1.6.0
Flask                              1.0.2
funcsigs                           1.0.2
grako                              3.99.9
html5lib                           1.0.1
idna                               2.8
ipaddress                          1.0.22
ipython                            7.4.0
ipython-genutils                   0.2.0
isodate                            0.6.0
itsdangerous                       1.1.0
jedi                               0.13.3
Jinja2                             2.10
jsonschema                         3.0.1
liac-arff                          2.4.0
lockfile                           0.12.2
lxml                               4.3.3
Mako                               1.0.8
MarkupSafe                         1.1.1
mock                               2.0.0
more-itertools                     7.0.0
netaddr                            0.7.19
netifaces                          0.10.9
networkx                           2.2
nose                               1.3.7
olefile                            0.46
packaging                          19.0
paramiko                           2.4.2
parso                              0.3.4
pathlib2                           2.3.3
paycheck                           1.0.2
pbr                                5.1.3
pexpect                            4.6.0
pickleshare                        0.7.5
Pillow                             6.0.0
pip                                18.1
pluggy                             0.9.0
ply                                3.11
prompt-toolkit                     2.0.9
protobuf                           3.7.1
prov                               1.5.3
psutil                             5.6.1
psycopg2                           2.7.7
ptyprocess                         0.6.0
py                                 1.8.0
pyasn1                             0.4.5
pycparser                          2.19
pycrypto                           2.6.1
pydot                              1.4.1
pydotplus                          2.0.2
Pygments                           2.3.1
PyNaCl                             1.3.0
pyparsing                          2.3.1
pyrsistent                         0.14.11
pytest                             4.4.0
pytest-forked                      1.0.2
pytest-runner                      4.4
pytest-xdist                       1.27.0
python-dateutil                    2.8.0
python-editor                      1.0.4
python-gflags                      3.1.2
python-oauth2                      1.1.0
pytz                               2018.9
PyYAML                             5.1
rdflib                             4.2.2
requests                           2.21.0
Rtree                              0.8.3
scandir                            1.10.0
setuptools                         40.6.2
setuptools-scm                     3.2.0
simplegeneric                      0.8.1
simplejson                         3.16.0
six                                1.12.0
SPARQLWrapper                      1.8.2
SQLAlchemy                         1.3.1
TatSu                              4.2.6
toolz                              0.9.0
traitlets                          4.3.2
traits                             5.0.0
typing                             3.6.6
urllib3                            1.24.1
vcversioner                        2.16.0.0
wcwidth                            0.1.7
webencodings                       0.5.1
Werkzeug                           0.15.1
wheel                              0.33.1
XlsxWriter                         1.1.5```

Running with an option
```]$ pip install --install-option='--vec' arbor
/p/software/jusuf/stages/2019a/software/Python/3.6.8-GCCcore-8.3.0/lib/python3.6/site-packages/pip/_internal/commands/install.py:211: UserWarning: Disabling all use of wheels due to the use of --build-options / --global-options / --install-options.
  cmdoptions.check_install_build_global(options)
Collecting arbor
  Using cached https://files.pythonhosted.org/packages/88/3a/2dfb1af9b3a8b32b227ef734ab152de299cd696bb7b59a710d5d4b232c5e/arbor-0.3.tar.gz
Skipping bdist_wheel for arbor, due to binaries being disabled for it.
Installing collected packages: arbor```

As no wheel is used, the error does not appear.

**Comment by bcumming** [comment]
We never use a wheel. This is an issue with some combinations of Python + Pip + Setuptools which skips the `install` step before `build_ext` when not building a wheel.
Passing an argument forces it to run the `install` step, which sets the `cl_opt_arch`, `cl_opt_mpi` etc variables.

Can you try again, after upgrading pip?
```
pip install --upgrade pip
pip install arbor
```
Then, if that doesn't fix the issue:
```
pip install --upgrade setuptools
pip install arbor
```

This is to check whether an old version of pip or setuptools is causing the issue.

**Comment by thorstenhater** [comment]
Upgrading to pip 20.0.2 fixes the issue.

**Comment by bcumming** [comment]
Thanks! This confirms my suspicion that older versions of pip don't run the `install` step if no arguments are passed when doing an external build, which is how the Arbor package is installed.

**Comment by thorstenhater** [comment]
I will close this then. Maybe leave a note that Arbor requires pip > 19?

**Comment by bcumming** [comment]
Let's keep it open until we have a fix.
I have to figure out how to enforce that requirement in a user friendly way.
We can close this when we have a PR merged with the solution.

**Comment by bcumming** [comment]
Fixed by #1010
